File: day-6/day6.pl
#!perl
use v5.24;
use warnings;

my $input = do { local ( @ARGV, $/ ) = 'input.txt'; <> };
say "Start of packet:  ", find_marker( \$input, 4 );
say "Start of message: ", find_marker( \$input, 14 );

sub find_marker {
	my ( $buffer, $marker_size ) = @_;
	for my $pos ( $marker_size .. length $$buffer ) {
		my %chars =
			map +( $_ => 1 ),
			unpack(
				'c' x $marker_size,
				substr( $$buffer, $pos - $marker_size, $marker_size ),
			);
		return $pos if keys(%chars) == $marker_size;
	}
	return;
}

File: day-6/day6.php
<?php

function find_marker ( &$buffer, $marker_size ) {
	$length = strlen( $buffer );
	for ( $pos = $marker_size; $pos <= $length; $pos++ ) {
		$chars = array_unique(
			str_split(
				substr( $buffer, $pos - $marker_size, $marker_size )
			)
		);
		if ( count($chars) == $marker_size )
			return $pos;
	}
	return FALSE;
}

$input = file_get_contents( "input.txt" );

echo 'Start of packet:  ' . find_marker( $input, 4 ) . "\n";
echo 'Start of message: ' . find_marker( $input, 14 ) . "\n";
